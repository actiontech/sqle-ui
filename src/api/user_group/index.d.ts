import {
  IGetUserGroupTipsResV1,
  IGetUserGroupsResV1,
  ICreateUserGroupReqV1,
  IBaseRes,
  IPatchUserGroupReqV1
} from '../common.d';

export interface IGetUserGroupTipListV1Return extends IGetUserGroupTipsResV1 {}

export interface IGetUserGroupListV1Params {
  filter_user_group_name?: string;

  page_index?: number;

  page_size?: number;
}

export interface IGetUserGroupListV1Return extends IGetUserGroupsResV1 {}

export interface ICreateUserGroupV1Params extends ICreateUserGroupReqV1 {}

export interface ICreateUserGroupV1Return extends IBaseRes {}

export interface IDeleteUserGroupV1Params {
  user_group_name: string;
}

export interface IDeleteUserGroupV1Return extends IBaseRes {}

export interface IUpdateUserGroupV1Params extends IPatchUserGroupReqV1 {
  user_group_name: string;
}

export interface IUpdateUserGroupV1Return extends IBaseRes {}
