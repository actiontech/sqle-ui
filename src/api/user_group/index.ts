/* tslint:disable no-identical-functions */
/* tslint:disable no-useless-cast */
/* tslint:disable no-unnecessary-type-assertion */
/* tslint:disable no-big-function  */
/* tslint:disable no-duplicate-string  */
import ServiceBase from '../Service.base';
import { AxiosRequestConfig } from 'axios';

import {
  IGetUserGroupTipListV1Return,
  IGetUserGroupListV1Params,
  IGetUserGroupListV1Return,
  ICreateUserGroupV1Params,
  ICreateUserGroupV1Return,
  IDeleteUserGroupV1Params,
  IDeleteUserGroupV1Return,
  IUpdateUserGroupV1Params,
  IUpdateUserGroupV1Return
} from './index.d';

class UserGroupService extends ServiceBase {
  public getUserGroupTipListV1(options?: AxiosRequestConfig) {
    return this.get<IGetUserGroupTipListV1Return>(
      '/v1/user_group_tips',
      undefined,
      options
    );
  }

  public getUserGroupListV1(
    params: IGetUserGroupListV1Params,
    options?: AxiosRequestConfig
  ) {
    const paramsData = this.cloneDeep(params);
    return this.get<IGetUserGroupListV1Return>(
      '/v1/user_groups',
      paramsData,
      options
    );
  }

  public CreateUserGroupV1(
    params: ICreateUserGroupV1Params,
    options?: AxiosRequestConfig
  ) {
    const paramsData = this.cloneDeep(params);
    return this.post<ICreateUserGroupV1Return>(
      '/v1/user_groups',
      paramsData,
      options
    );
  }

  public deleteUserGroupV1(
    params: IDeleteUserGroupV1Params,
    options?: AxiosRequestConfig
  ) {
    const paramsData = this.cloneDeep(params);
    const user_group_name = paramsData.user_group_name;
    delete paramsData.user_group_name;

    return this.delete<IDeleteUserGroupV1Return>(
      `/v1/user_groups/${user_group_name}/`,
      paramsData,
      options
    );
  }

  public updateUserGroupV1(
    params: IUpdateUserGroupV1Params,
    options?: AxiosRequestConfig
  ) {
    const paramsData = this.cloneDeep(params);
    const user_group_name = paramsData.user_group_name;
    delete paramsData.user_group_name;

    return this.patch<IUpdateUserGroupV1Return>(
      `/v1/user_groups/${user_group_name}/`,
      paramsData,
      options
    );
  }
}

export default new UserGroupService();
