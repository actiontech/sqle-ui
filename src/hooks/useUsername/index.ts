import { useBoolean } from 'ahooks';
import React from 'react';
import { IUserTipResV1 } from '../../api/common';
import user from '../../api/user';
import { ResponseCode } from '../../data/common';

const useUsername = () => {
  const [usernameList, setUsernameList] = React.useState<IUserTipResV1[]>([]);
  const [loading, { setTrue, setFalse }] = useBoolean();

  const updateUsernameList = React.useCallback(() => {
    setTrue();
    user
      .getUserTipListV1()
      .then((res) => {
        if (res.data.code === ResponseCode.SUCCESS) {
          setUsernameList(res.data?.data ?? []);
        } else {
          setUsernameList([]);
        }
      })
      .catch(() => {
        setUsernameList([]);
      })
      .finally(() => {
        setFalse();
      });
  }, [setFalse, setTrue]);

  return {
    usernameList,
    loading,
    updateUsernameList,
  };
};

export default useUsername;
