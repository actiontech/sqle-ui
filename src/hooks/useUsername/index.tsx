import { useBoolean } from 'ahooks';
import { Select } from 'antd';
import React from 'react';
import { IUserTipResV1 } from '../../api/common';
import user from '../../api/user';
import { ResponseCode } from '../../data/common';

const useUsername = () => {
  const [usernameList, setUsernameList] = React.useState<IUserTipResV1[]>([]);
  const [loading, { setTrue, setFalse }] = useBoolean();

  const updateUsernameList = React.useCallback(() => {
    setTrue();
    user
      .getUserTipListV1({})
      .then((res) => {
        if (res.data.code === ResponseCode.SUCCESS) {
          setUsernameList(res.data?.data ?? []);
        } else {
          setUsernameList([]);
        }
      })
      .catch(() => {
        setUsernameList([]);
      })
      .finally(() => {
        setFalse();
      });
  }, [setFalse, setTrue]);

  const generateUsernameSelectOption = React.useCallback(() => {
    return usernameList.map((user) => {
      return (
        <Select.Option key={user.user_name} value={user.user_name ?? ''}>
          {user.user_name}
        </Select.Option>
      );
    });
  }, [usernameList]);

  return {
    usernameList,
    loading,
    updateUsernameList,
    generateUsernameSelectOption,
  };
};

export default useUsername;
