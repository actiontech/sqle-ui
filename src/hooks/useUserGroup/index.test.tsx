import { act, renderHook } from '@testing-library/react-hooks';
import {
  render,
  fireEvent,
  screen,
  act as reactAct,
  cleanup,
} from '@testing-library/react';
import {
  rejectThreeSecond,
  resolveErrorThreeSecond,
  resolveThreeSecond,
} from '../../testUtils/mockRequest';
import { Select } from 'antd';
import useUsername from '.';
import user_group from '../../api/user_group';

describe('useUserGroup', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  const mockRequest = () => {
    const spy = jest.spyOn(user_group, 'getUserGroupTipListV1');
    return spy;
  };

  test('should get group data from request', async () => {
    const requestSpy = mockRequest();
    requestSpy.mockImplementation(() =>
      resolveThreeSecond([{ user_group_name: 'group1' }])
    );
    const { result, waitForNextUpdate } = renderHook(() => useUsername());
    expect(result.current.loading).toBe(false);
    expect(result.current.userGroupList).toEqual([]);
    const { baseElement } = render(
      <Select>{result.current.generateUserGroupSelectOption()}</Select>
    );
    expect(baseElement).toMatchSnapshot();

    act(() => {
      result.current.updateUserGroupList('test');
    });

    expect(result.current.loading).toBe(true);
    expect(requestSpy).toBeCalledTimes(1);
    expect(requestSpy).toBeCalledWith({
      filter_project: 'test',
    });
    expect(result.current.userGroupList).toEqual([]);

    jest.advanceTimersByTime(3000);
    await waitForNextUpdate();

    expect(result.current.loading).toBe(false);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.userGroupList).toEqual([
      { user_group_name: 'group1' },
    ]);
    cleanup();

    const { baseElement: baseElementWithOptions } = render(
      <Select data-testid="testId" value="value1">
        {result.current.generateUserGroupSelectOption()}
      </Select>
    );
    expect(baseElementWithOptions).toMatchSnapshot();

    reactAct(() => {
      fireEvent.mouseDown(screen.getByText('value1'));
      jest.runAllTimers();
    });

    await screen.findAllByText('group1');
    expect(baseElementWithOptions).toMatchSnapshot();
  });

  test('should set list to empty array when response code is not equal success code', async () => {
    const requestSpy = mockRequest();
    requestSpy.mockImplementation(() =>
      resolveThreeSecond([{ user_group_name: 'group1' }])
    );
    const { result, waitForNextUpdate } = renderHook(() => useUsername());
    expect(result.current.loading).toBe(false);
    expect(result.current.userGroupList).toEqual([]);

    act(() => {
      result.current.updateUserGroupList();
    });

    expect(result.current.loading).toBe(true);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.userGroupList).toEqual([]);

    jest.advanceTimersByTime(3000);
    await waitForNextUpdate();

    expect(result.current.loading).toBe(false);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.userGroupList).toEqual([
      { user_group_name: 'group1' },
    ]);
    requestSpy.mockClear();
    requestSpy.mockImplementation(() =>
      resolveErrorThreeSecond([{ user_group_name: 'group1' }])
    );

    act(() => {
      result.current.updateUserGroupList();
    });
    expect(result.current.loading).toBe(true);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.userGroupList).toEqual([
      {
        user_group_name: 'group1',
      },
    ]);

    jest.advanceTimersByTime(3000);
    await waitForNextUpdate();

    expect(result.current.loading).toBe(false);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.userGroupList).toEqual([]);
  });

  test('should set list to empty array when response throw error', async () => {
    const requestSpy = mockRequest();
    requestSpy.mockImplementation(() =>
      resolveThreeSecond([{ user_group_name: 'group1' }])
    );
    const { result, waitForNextUpdate } = renderHook(() => useUsername());
    expect(result.current.loading).toBe(false);
    expect(result.current.userGroupList).toEqual([]);

    act(() => {
      result.current.updateUserGroupList();
    });

    expect(result.current.loading).toBe(true);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.userGroupList).toEqual([]);

    jest.advanceTimersByTime(3000);
    await waitForNextUpdate();

    expect(result.current.loading).toBe(false);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.userGroupList).toEqual([
      { user_group_name: 'group1' },
    ]);
    requestSpy.mockClear();
    requestSpy.mockImplementation(() =>
      rejectThreeSecond([{ user_group_name: 'group1' }])
    );

    act(() => {
      result.current.updateUserGroupList();
    });
    expect(result.current.loading).toBe(true);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.userGroupList).toEqual([
      {
        user_group_name: 'group1',
      },
    ]);

    jest.advanceTimersByTime(3000);
    await waitForNextUpdate();

    expect(result.current.loading).toBe(false);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.userGroupList).toEqual([]);
  });
});
