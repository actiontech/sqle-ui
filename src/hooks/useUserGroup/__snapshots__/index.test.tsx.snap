// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useUserGroup should get group data from request 1`] = `
<body>
  <div>
    <div
      class="ant-select ant-select-single ant-select-show-arrow"
    >
      <div
        class="ant-select-selector"
      >
        <span
          class="ant-select-selection-search"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="ant-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
        </span>
        <span
          class="ant-select-selection-placeholder"
        />
      </div>
      <span
        aria-hidden="true"
        class="ant-select-arrow"
        style="user-select: none;"
        unselectable="on"
      >
        <span
          aria-label="down"
          class="anticon anticon-down ant-select-suffix"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="down"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
            />
          </svg>
        </span>
      </span>
    </div>
  </div>
</body>
`;

exports[`useUserGroup should get group data from request 2`] = `
<body>
  <div>
    <div
      class="ant-select ant-select-single ant-select-show-arrow"
      data-testid="testId"
    >
      <div
        class="ant-select-selector"
      >
        <span
          class="ant-select-selection-search"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="ant-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
        </span>
        <span
          class="ant-select-selection-item"
          title="value1"
        >
          value1
        </span>
      </div>
      <span
        aria-hidden="true"
        class="ant-select-arrow"
        style="user-select: none;"
        unselectable="on"
      >
        <span
          aria-label="down"
          class="anticon anticon-down ant-select-suffix"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="down"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
            />
          </svg>
        </span>
      </span>
    </div>
  </div>
</body>
`;

exports[`useUserGroup should get group data from request 3`] = `
<body>
  <div>
    <div
      class="ant-select ant-select-single ant-select-show-arrow ant-select-open"
      data-testid="testId"
    >
      <div
        class="ant-select-selector"
      >
        <span
          class="ant-select-selection-search"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-expanded="true"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="ant-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
        </span>
        <span
          class="ant-select-selection-item"
          title="value1"
        >
          value1
        </span>
      </div>
      <span
        aria-hidden="true"
        class="ant-select-arrow"
        style="user-select: none;"
        unselectable="on"
      >
        <span
          aria-label="down"
          class="anticon anticon-down ant-select-suffix"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="down"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
            />
          </svg>
        </span>
      </span>
    </div>
  </div>
  <div
    style="position: absolute; top: 0px; left: 0px; width: 100%;"
  >
    <div>
      <div
        class="ant-select-dropdown"
        style="opacity: 0; min-width: 0; width: 0px;"
      >
        <div>
          <div
            id="rc_select_TEST_OR_SSR_list"
            role="listbox"
            style="height: 0px; width: 0px; overflow: hidden;"
          >
            <div
              aria-selected="false"
              id="rc_select_TEST_OR_SSR_list_0"
              role="option"
            >
              group1
            </div>
          </div>
          <div
            class="rc-virtual-list"
            style="position: relative;"
          >
            <div
              class="rc-virtual-list-holder"
              style="max-height: 256px; overflow-y: hidden;"
            >
              <div>
                <div
                  class="rc-virtual-list-holder-inner"
                  style="display: flex; flex-direction: column;"
                >
                  <div
                    aria-selected="false"
                    class="ant-select-item ant-select-item-option ant-select-item-option-active"
                  >
                    <div
                      class="ant-select-item-option-content"
                    >
                      group1
                    </div>
                    <span
                      aria-hidden="true"
                      class="ant-select-item-option-state"
                      style="user-select: none;"
                      unselectable="on"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="rc-virtual-list-scrollbar"
              style="width: 8px; top: 0px; bottom: 0px; right: 0px; position: absolute; display: none;"
            >
              <div
                class="rc-virtual-list-scrollbar-thumb"
                style="width: 100%; height: 128px; top: 0px; left: 0px; position: absolute; background: rgba(0, 0, 0, 0.5); border-radius: 99px; cursor: pointer; user-select: none;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`useUserGroup should match snapshot when need show tooltip when generate select options 1`] = `
Array [
  <Option
    value="group1"
  >
    <EmptyBox
      defaultNode="group1"
      if={true}
    >
      <Tooltip
        placement="right"
        title={
          <React.Fragment>
            user_name0, user_name1, user_name2, user_name3, user_name4, user_name5, user_name6, user_name7, user_name8, user_name9, user_name10, user_name11, user_name12, user_name13, user_name14
            ...
          </React.Fragment>
        }
      >
        <div
          className="full-width-element"
        >
          group1
        </div>
      </Tooltip>
    </EmptyBox>
  </Option>,
]
`;

exports[`useUserGroup should match snapshot when need show tooltip when generate select options 2`] = `
Array [
  <Option
    value="group1"
  >
    <EmptyBox
      defaultNode="group1"
      if={true}
    >
      <Tooltip
        placement="right"
        title={
          <React.Fragment>
            user_name0, user_name1, user_name2, user_name3, user_name4, user_name5, user_name6, user_name7, user_name8, user_name9
            
          </React.Fragment>
        }
      >
        <div
          className="full-width-element"
        >
          group1
        </div>
      </Tooltip>
    </EmptyBox>
  </Option>,
]
`;
